File: algorithms/medium/132_patten.ex
defmodule Solution do
  @spec find132pattern(nums :: [integer]) :: boolean
  def find132pattern([n | ns]) do
    ns
    |> Enum.reduce_while({[], n}, fn n, {st, cur_min} ->
      st = Enum.drop_while(st, fn {m, _} -> m <= n end)

      cond do
        st == [] -> {:cont, {[{n, cur_min} | st], min(cur_min, n)}}
        st |> hd() |> elem(1) < n -> {:halt, :ok}
        true -> {:cont, {[{n, cur_min} | st], min(cur_min, n)}}
      end
    end)
    |> then(&(&1 == :ok))
  end
end

File: algorithms/medium/132_patten.rs
impl Solution {
    pub fn find132pattern(nums: Vec<i32>) -> bool {
        let mut st: Vec<(i32, i32)> = Vec::new();
        let mut cur_min = nums[0];
        for &n in &nums[1..] {
            while !st.is_empty() && st[st.len() - 1].0 <= n {
                st.pop();
            }
            if !st.is_empty() && st[st.len() - 1].1 < n {
                return true;
            }
            st.push((n, cur_min));
            cur_min = cur_min.min(n);
        }
        
        false
    }
}
